### PR TITLE
JDK-8314056 Remove runtime platform check from frem/drem

### PR DESCRIPTION
**Pull request by asgibbons**
Remove platform check and move code to stubGenerator.  This fix increases performance by ~4.5%.

UPDATE: Subsequent commits increase performance gain to ~2x for AVX2, with no significant change to AVX512.

Tested tier1.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314056](https://bugs.openjdk.org/browse/JDK-8314056): Remove runtime platform check from frem/drem (**Enhancement** - P4)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15210/head:pull/15210` \
`$ git checkout pull/15210`

Update a local copy of the PR: \
`$ git checkout pull/15210` \
`$ git pull https://git.openjdk.org/jdk.git pull/15210/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15210`

View PR using the GUI difftool: \
`$ git pr show -t 15210`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15210.diff">https://git.openjdk.org/jdk/pull/15210.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15210#issuecomment-1671831668)

### CODE REVIEW

File: src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp [hunk 1]
```diff
@@ -292,11 +292,6 @@ address StubGenerator::generate_libmFmod() {
     //   __asm { stmxcsr DWORD PTR[mxcsr] }
     //   mxcsr_rz = 0x7f80 | mxcsr;
     __ push(rax);
-    __ stmxcsr(Address(rsp, 0));
-    __ movl(rax, Address(rsp, 0));
-    __ movl(rcx, rax);
-    __ orl(rcx, 0x7f80);
-    __ movl(Address(rsp, 0x04), rcx);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp:299]: No need to do stmxcsr, we could define mxcsr_rz on similar lines as mxcsr_std as part of StubGenerator::create_control_words() and use those in fmod.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp [hunk 2]
```diff
@@ -319,6 +314,12 @@ address StubGenerator::generate_libmFmod() {
...
     __ bind(L_104a);
+    __ stmxcsr(Address(rsp, 0));
+    __ movl(rax, Address(rsp, 0));
+    __ movl(rcx, rax);
+    __ orl(rcx, 0x7f80);
+    __ movl(Address(rsp, 0x04), rcx);
+
     __ andl(rax, 0x6000);
     __ cmpl(rax, 0x2000);
     __ jcc(Assembler::equal, L_10c1);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp:325]: Some of the jcc and jmp could be generated as short branches where feasible.
Also the stub and the loop entries could be aligned to 16 bytes for better perf.

**Comment by sviswa7** [comment]
The changes look good to me now. Could you please also get the performance of DremFrem jmh micro with/without your PR for both UseAVX=2 and UseAVX=3?

**Comment by asgibbons** [comment]
> The changes look good to me now. Could you please also get the performance of DremFrem jmh micro with/without your PR for both UseAVX=2 and UseAVX=3?

Results of DremFrem:

Before these changes:
```
java -XX:UseAVX=2 -jar ./build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 1 DremFrem:

Benchmark                      Mode  Cnt  Score   Error  Units
DremFrem.calcDoubleJava        avgt    5  5.645 ± 0.003  ns/op
DremFrem.calcFloatJava         avgt    5  6.044 ± 0.005  ns/op
DremFrem.cornercaseDoubleJava  avgt    5  5.462 ± 0.033  ns/op
DremFrem.cornercaseFloatJava   avgt    5  5.494 ± 0.045  ns/op

java -XX:UseAVX=3 -jar ./build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 1 DremFrem:

Benchmark                      Mode  Cnt  Score   Error  Units
DremFrem.calcDoubleJava        avgt    5  2.916 ± 0.006  ns/op
DremFrem.calcFloatJava         avgt    5  4.016 ± 0.001  ns/op
DremFrem.cornercaseDoubleJava  avgt    5  5.506 ± 0.009  ns/op
DremFrem.cornercaseFloatJava   avgt    5  5.506 ± 0.010  ns/op

```
=============================================
WITH these changes:

```
java -XX:UseAVX=2 -jar ./build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 1 DremFrem:

Benchmark                      Mode  Cnt  Score   Error  Units
DremFrem.calcDoubleJava        avgt    5  2.838 ± 0.001  ns/op
DremFrem.calcFloatJava         avgt    5  3.554 ± 0.002  ns/op
DremFrem.cornercaseDoubleJava  avgt    5  5.511 ± 0.022  ns/op
DremFrem.cornercaseFloatJava   avgt    5  5.467 ± 0.014  ns/op

java -XX:UseAVX=3 -jar ./build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 1 DremFrem:

Benchmark                      Mode  Cnt  Score    Error  Units
DremFrem.calcDoubleJava        avgt    5  2.598 ±  0.001  ns/op
DremFrem.calcFloatJava         avgt    5  3.229 ±  0.001  ns/op
DremFrem.cornercaseDoubleJava  avgt    5  5.508 ±  0.014  ns/op
DremFrem.cornercaseFloatJava   avgt    5  5.511 ±  0.005  ns/op

```

**Review by sviswa7 — APPROVED** [review]
The PR looks good to me. Thanks for considering all the review comments.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp [hunk 3]
```diff
@@ -480,9 +470,9 @@ address StubGenerator::generate_libmFmod() {
     //   while (bs <= a)
     __ ucomisd(xmm0, xmm1);
     __ movapd(xmm4, xmm0);
-    __ jcc(Assembler::aboveEqual, L_117f);
-    __ jmp(L_11af);
-    __ align32();
+    __ jccb(Assembler::aboveEqual, L_117f);
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp:473]: This is a bunded label jump, Assembler::jcc should automatically optimize it with short jump encoding, but its not a blocker.

**Review by jatin-bhateja — COMMENTED** [review]
Changes looks good to me, you may also include the results of following benchmark.
./test/micro/org/openjdk/bench/vm/compiler/pea/Blender.java

**Comment by asgibbons** [comment]
@jddarcy FYI.  If you'd like to review, it would be appreciated.  Thanks.

**Review by jatin-bhateja — APPROVED** [review]

**Pull request closed without merging**